From b68694436cf068146b8d8ba43fafc418a32de0f4 Mon Sep 17 00:00:00 2001
From: John Reck <jreck@google.com>
Date: Thu, 7 Aug 2025 16:51:23 -0400
Subject: [PATCH] Backport DNG SDK 1.7.1 support in SkRawCodec

Bug: 412662901
Test: make, otherwise clean cherry-pick of main's version

Changes coming from ag/34978263

Merged-In: I800bef8685a915187a897df09e37875093fc9af7

Change-Id: I735f93881f0f60db2057af74d8fd03492d7c9895
---
 Android.bp               | 10 +++++++++-
 src/codec/SkRawCodec.cpp | 14 ++++++++++----
 2 files changed, 19 insertions(+), 5 deletions(-)

diff --git a/Android.bp b/Android.bp
index 747e86ea5a..0374d64d69 100644
--- a/Android.bp
+++ b/Android.bp
@@ -1889,7 +1889,6 @@ cc_defaults {
     name: "skia_deps",
     defaults: ["skia_renderengine_deps"],
     shared_libs: [
-        "libdng_sdk",
         "libjpeg",
         "libpiex",
         "libexpat",
@@ -1905,12 +1904,21 @@ cc_defaults {
       android: {
         shared_libs: [
             "libheif",
+            "libdng_sdk",
+        ],
+      },
+      host_linux: {
+        static_libs: [
+          "libdng_sdk",
         ],
       },
       darwin: {
         host_ldlibs: [
             "-framework AppKit",
         ],
+        static_libs: [
+          "libdng_sdk",
+        ],
       },
       windows: {
         host_ldlibs: [
diff --git a/src/codec/SkRawCodec.cpp b/src/codec/SkRawCodec.cpp
index c040418ddb..33d9948238 100644
--- a/src/codec/SkRawCodec.cpp
+++ b/src/codec/SkRawCodec.cpp
@@ -123,7 +123,7 @@ class SkDngHost : public dng_host {
 public:
     explicit SkDngHost(dng_memory_allocator* allocater) : dng_host(allocater) {}
 
-    void PerformAreaTask(dng_area_task& task, const dng_rect& area) override {
+    void PerformAreaTask(dng_area_task& task, const dng_rect& area, dng_area_task_progress*) override {
         SkTaskGroup taskGroup;
 
         // tileSize is typically 256x256
@@ -134,11 +134,11 @@ public:
 
         SkMutex mutex;
         SkTArray<dng_exception> exceptions;
-        task.Start(numTasks, tileSize, &Allocator(), Sniffer());
+        task.Start(numTasks, area, tileSize, &Allocator(), Sniffer());
         for (int taskIndex = 0; taskIndex < numTasks; ++taskIndex) {
             taskGroup.add([&mutex, &exceptions, &task, this, taskIndex, taskAreas, tileSize] {
                 try {
-                    task.ProcessOnThread(taskIndex, taskAreas[taskIndex], tileSize, this->Sniffer());
+                    task.ProcessOnThread(taskIndex, taskAreas[taskIndex], tileSize, this->Sniffer(), nullptr);
                 } catch (dng_exception& exception) {
                     SkAutoMutexExclusive lock(mutex);
                     exceptions.push_back(exception);
@@ -434,7 +434,13 @@ private:
 class SkDngStream : public dng_stream {
 public:
     // Will NOT take the ownership of the stream.
-    SkDngStream(SkRawStream* stream) : fStream(stream) {}
+    SkDngStream(SkRawStream* stream)
+        // The default constructor sets offsetInOriginalFile to invalid (-1), however
+        // this results in dng_negative hitting a bug path that causes a crash due to
+        // unsigned overflow occuring. This offset doesn't seem to serve any purpose,
+        // so just pretend we're always at the start of the file to avoid the crash
+        : dng_stream((dng_abort_sniffer*) nullptr, dng_stream::kDefaultBufferSize, 0)
+        , fStream(stream) {}
 
     ~SkDngStream() override {}
 
-- 
2.52.0.rc1.455.g30608eb744-goog

